File: android/src/main/java/io/appmetrica/analytics/flutter/impl/ReporterImpl.java
package io.appmetrica.analytics.flutter.impl;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import io.appmetrica.analytics.AppMetrica;
import io.appmetrica.analytics.ecommerce.ECommerceEvent;
import io.appmetrica.analytics.flutter.pigeon.Pigeon;
import io.appmetrica.analytics.flutter.utils.Converter;
import io.appmetrica.analytics.flutter.utils.ECommerceConverter;
import io.appmetrica.analytics.plugins.PluginErrorDetails;

public class ReporterImpl implements Pigeon.ReporterPigeon {

    @NonNull
    final Context context;

    public ReporterImpl(@NonNull final Context context) {
        this.context = context;
    }

    @Override
    public void sendEventsBuffer(@NonNull String apiKey) {
        AppMetrica.getReporter(context, apiKey).sendEventsBuffer();
    }

    @Override
    public void setDataSendingEnabled(@NonNull String apiKey, @NonNull Boolean enabled) {
        AppMetrica.getReporter(context, apiKey).setDataSendingEnabled(enabled);
    }

    @Override
    public void reportEvent(@NonNull String apiKey, @NonNull String eventName) {
        AppMetrica.getReporter(context, apiKey).reportEvent(eventName);
    }

    @Override
    public void reportEventWithJson(@NonNull String apiKey, @NonNull String eventName, @Nullable String attributesJson) {
        AppMetrica.getReporter(context, apiKey).reportEvent(eventName, attributesJson);
    }

    @Override
    public void reportError(@NonNull String apiKey, @NonNull Pigeon.ErrorDetailsPigeon error, @Nullable String message) {
        AppMetrica.getReporter(context, apiKey).getPluginExtension()
            .reportError(Converter.toNative(error), message);
    }

    @Override
    public void reportErrorWithGroup(
        @NonNull String apiKey,
        @NonNull String groupId,
        @Nullable Pigeon.ErrorDetailsPigeon error,
        @Nullable String message
    ) {
        AppMetrica.getReporter(context, apiKey).getPluginExtension()
            .reportError(groupId, message, error != null ? Converter.toNative(error) : null);
    }

    @Override
    public void reportUnhandledException(@NonNull String apiKey, @NonNull Pigeon.ErrorDetailsPigeon error) {
        AppMetrica.getReporter(context, apiKey).getPluginExtension()
            .reportUnhandledException(Converter.toNative(error));
    }

    @Override
    public void resumeSession(@NonNull String apiKey) {
        AppMetrica.getReporter(context, apiKey).resumeSession();
    }

    @Override
    public void clearAppEnvironment(@NonNull String apiKey) {
        AppMetrica.getReporter(context, apiKey).clearAppEnvironment();
    }

    @Override
    public void pauseSession(@NonNull String apiKey) {
        AppMetrica.getReporter(context, apiKey).pauseSession();
    }

    @Override
    public void putAppEnvironmentValue(@NonNull String apiKey, @NonNull String key, @Nullable String value) {
        AppMetrica.getReporter(context, apiKey).putAppEnvironmentValue(key, value);
    }

    @Override
    public void setUserProfileID(@NonNull String apiKey, @Nullable String userProfileID) {
        AppMetrica.getReporter(context, apiKey).setUserProfileID(userProfileID);
    }

    @Override
    public void reportUserProfile(@NonNull String apiKey, @NonNull Pigeon.UserProfilePigeon userProfile) {
        AppMetrica.getReporter(context, apiKey).reportUserProfile(Converter.toNative(userProfile));
    }

    @Override
    public void reportRevenue(@NonNull String apiKey, @NonNull Pigeon.RevenuePigeon revenue) {
        AppMetrica.getReporter(context, apiKey).reportRevenue(Converter.toNative(revenue));
    }

    @Override
    public void reportECommerce(@NonNull String apiKey, @NonNull Pigeon.ECommerceEventPigeon event) {
        final ECommerceEvent eCommerceEvent = ECommerceConverter.toNative(event);
        if (eCommerceEvent != null) {
            AppMetrica.getReporter(context, apiKey).reportECommerce(eCommerceEvent);
        }
    }

    @Override
    public void reportAdRevenue(@NonNull String apiKey, @NonNull Pigeon.AdRevenuePigeon adRevenue) {
        AppMetrica.getReporter(context, apiKey).reportAdRevenue(Converter.toNative(adRevenue));
    }
}

File: lib/src/reporter/reporter_impl.dart
import 'dart:convert';

import 'package:appmetrica_plugin/src/ecommerce_event.dart';

import '../ad_revenue.dart';
import '../error_description.dart';
import '../appmetrica_api_pigeon.dart';
import '../pigeon_converter.dart';
import '../profile/attribute.dart';
import '../revenue.dart';
import 'reporter.dart';

class ReporterImpl implements AppMetricaReporter {
  static final _reporterPigeon = ReporterPigeon();

  final String _apiKey;

  ReporterImpl(this._apiKey);

  @override
  Future<void> clearAppEnvironment() =>
      _reporterPigeon.clearAppEnvironment(_apiKey);

  @override
  Future<void> pauseSession() =>
      _reporterPigeon.pauseSession(_apiKey);

  @override
  Future<void> putAppEnvironmentValue(String key, String? value) =>
      _reporterPigeon.putAppEnvironmentValue(_apiKey, key, value);

  @override
  Future<void> reportAdRevenue(AppMetricaAdRevenue adRevenue) =>
      _reporterPigeon.reportAdRevenue(_apiKey, adRevenue.toPigeon());

  @override
  Future<void> reportECommerce(AppMetricaECommerceEvent event) =>
      _reporterPigeon.reportECommerce(_apiKey, event.toPigeon());

  @override
  Future<void> reportError({
    String? message,
    AppMetricaErrorDescription? errorDescription
  }) =>
      _reporterPigeon.reportError(_apiKey, errorDescription.tryToAddCurrentTrace().toPigeon(), message);

  @override
  Future<void> reportErrorWithGroup(
      String groupId, {
        AppMetricaErrorDescription? errorDescription,
        String? message
      }) =>
      _reporterPigeon.reportErrorWithGroup(_apiKey, groupId, errorDescription?.toPigeon(), message);

  @override
  Future<void> reportEvent(String eventName) =>
      _reporterPigeon.reportEvent(_apiKey, eventName);

  @override
  Future<void> reportEventWithJson(String eventName, String? attributesJson) =>
      _reporterPigeon.reportEventWithJson(_apiKey, eventName, attributesJson);

  @override
  Future<void> reportEventWithMap(
      String eventName,
      Map<String, Object>? attributes
  ) =>
      _reporterPigeon.reportEventWithJson(_apiKey, eventName, jsonEncode(attributes));

  @override
  Future<void> reportRevenue(AppMetricaRevenue revenue) =>
      _reporterPigeon.reportRevenue(_apiKey, revenue.toPigeon());

  @override
  Future<void> reportUnhandledException(AppMetricaErrorDescription error) =>
      _reporterPigeon.reportUnhandledException(_apiKey, error.toPigeon());

  @override
  Future<void> reportUserProfile(AppMetricaUserProfile userProfile) =>
      _reporterPigeon.reportUserProfile(_apiKey, userProfile.toPigeon());

  @override
  Future<void> resumeSession() =>
      _reporterPigeon.resumeSession(_apiKey);

  @override
  Future<void> sendEventsBuffer() =>
      _reporterPigeon.sendEventsBuffer(_apiKey);

  @override
  Future<void> setDataSendingEnabled(bool enabled) =>
      _reporterPigeon.setDataSendingEnabled(_apiKey, enabled);

  @override
  Future<void> setUserProfileID(String? userProfileID) =>
      _reporterPigeon.setUserProfileID(_apiKey, userProfileID);
}
